activesupport/lib/active_support/cache/key_master.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Cache
+    # This class is responsible for cohercing input into a valid
+    # key format and cache version for cache stores.
+    #
+    # The goal of this class is to provide a composable way
+    # to build keys. You can put an instance of KeyMaster
+    # into another instance of KeyMaster and it should produce
+    # the same `cache_key` and `cache_version`.
+    #
+    # Example:
+    #
+    #   k1 = KeyMaster.new("i_aint_afraid_of_no_ghosts")
+    #   k2 = KeyMaster.new(k1)
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method
+    # key fragments can be added onto an existing key through the
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = KeyMaster.new
+    #   key << "who-ya-gonna-call"
+    #   key << "ghost-busters"
+    #   key.cache_key
+    #   # => "who-ya-gonna-call/ghost-busters"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = KeyMaster.new("slimer")
+    #   k1 << "stay-puft"
+    #   k2 = KeyMaster.new(["slimer", "stay-puft"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = KeyMaster.new(["slimer", ["stay-puft"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # KeyMaster object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   KeyMaster.new("there").cache_key
+    #   # => "there"
+    #
+    #   KeyMaster.new(["is", "no"]).cache_key
+    #   # => "is/no"
+    #
+    #   KeyMaster.new({ dana: 1, only: 2 }).cache_key
+    #   # => "dana=1/only=2"
+    #
+    #   klass = Class.new { def cache_key; "zuul"; end }
+    #   KeyMaster.new(klass.new).cache_key
+    #   # => "zuul"
+    #
+    # When generating a a cache key, `nil` will be filtered
+    # out, so `["ghost", nil "busters"]` will generate the
+    # same output as ["ghost", "busters"]. If presence of a
+    # `nil` is important, then convert it to a string.
```

Comment:
Is this a behavior change? `nil` is typically meaningful. Removing it causes collisions.